File: examples/Java/src/main/java/com/imatix/zguide/ex1_7/WuClient.java
package com.imatix.zguide.ex1_7;

import com.imatix.zguide.commons.ZConstants;
import org.zeromq.ZMQ.Context;
import org.zeromq.ZMQ.Socket;

import java.util.StringTokenizer;

import static java.lang.String.format;
import static org.zeromq.ZMQ.SUB;
import static org.zeromq.ZMQ.context;

/**
 * Weather update client in Java
 * Connects SUB socket to tcp://localhost:5556
 * Collects weather updates and finds avg temp in zipcode
 *
 * @since 1.0
 */

public class WuClient {

    public static void main(String[] args) {
        Context context = context(1);
        System.out.println("Collecting updates from weather update server...");
        //  Socket to talk to server
        Socket subscriber = context.socket(SUB);
        subscriber.connect(ZConstants.URL.TCP_5556);
        //  Subscribe to zipCode, default is NYC, 10001
        String filter = (args.length > 0) ? args[0] : "10001";
        subscriber.subscribe(filter.getBytes());
        // Process 100 updates
        long totalTemp = 0;
        int updateNum = 0;
        while (updateNum < 100) {
            //  Use trim to remove the tailing '0' character
            String receivedStr = subscriber.recvStr(0).trim();
            StringTokenizer tokenizer = new StringTokenizer(receivedStr, " ");
            int zipCode = Integer.valueOf(tokenizer.nextToken());
            int temperature = Integer.valueOf(tokenizer.nextToken());
            int relHumidity = Integer.valueOf(tokenizer.nextToken());
            System.out.println(format("Weather update: zipCode: %d , temperature: %d , humidity: %d",
                    zipCode, temperature, relHumidity));
            totalTemp += temperature;
            updateNum++;
        }
        System.out.printf("Average Temperature for zipCode %s was %d", filter, (int) (totalTemp / updateNum));
        subscriber.close();
        context.term();
    }
}

File: examples/Free Pascal/wuclient.pas
{
  Weather update client
  Connects SUB socket to tcp://localhost:5556
  Collects weather updates and finds avg temp in zipcode
  @author cpicanco <cpicanco@ufpa.br>
}
program wuclient;

{$MODE objfpc}{$H+}{$COPERATORS ON}

uses SysUtils, zmq, zmq.helpers;

var
  context : Pointer;
  subscriber : Pointer;
  rc: Integer;
  filter : string;

  update_nbr : Integer;
  total_temp : LongInt = 0;
  total_humid : LongInt = 0;

  LString : shortstring;
  zipcode, temperature, relhumidity : integer;
begin
  //  Socket to talk to server
  WriteLn('Collecting updates from weather server…');
  context := zmq_ctx_new;
  subscriber := zmq_socket(context, ZMQ_SUB);
  rc := zmq_connect(subscriber, 'tcp://localhost:5556');
  Assert(rc = 0);

  //  Subscribe to zipcode, default is NYC, 10001
  if ParamCount > 1 then filter := ParamStr(1) else filter := '10001';
  rc := zmq_setsockopt(subscriber, ZMQ_SUBSCRIBE, @filter[1], Length(filter));
  Assert(rc = 0);

  //  Process 100 updates
  for update_nbr := 0 to 100 -1 do
    begin
      LString := s_recv(subscriber);
      ReadStr(LString, zipcode, temperature, relhumidity);
      total_temp += temperature;
      total_humid += relhumidity;
      LString := '';
    end;
  update_nbr += 1;

  WriteLn(Format('Average temperature for zipcode %s was %.2f',
    [filter, total_temp/update_nbr]));

  WriteLn(Format('Average relative humidity for zipcode %s was %.2f',
    [filter, total_humid/update_nbr]));

  zmq_close(subscriber);
  zmq_ctx_destroy(context);
end.

